File: implementations/csharp/vanilla/parser/ast/NodeList.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections;
using Parser.Ast;

namespace Parser.Ast
{
    /// <summary>
    /// Class which holds a list of nodes of type T
    /// </summary>
    public class NodeList : List<ISyntaxNode>, ISyntaxNode
    {
        #region Private Members

        private List<ISyntaxNode> list = new List<ISyntaxNode>();

        #endregion

        #region Public Members

        public new int Count = 0;

        #endregion

        #region Public Methods

        /// <summary>
        /// Add element to NodeList
        /// </summary>
        /// <param name="element">Element to add</param>
        public new void Add(ISyntaxNode element)
        {
            list.Add(element);
            Count++;
        }


        /// <summary>
        /// Get element from specified index
        /// </summary>
        /// <param name="index">Index of element to retrieve</param>
        /// <returns>Element</returns>
        public ISyntaxNode Get(int index)
        {
            return list.ElementAt(index);
        }

        /// <summary>
        /// Remove element at specified index
        /// </summary>
        /// <param name="index">Index of item to remove</param>
        public void Remove(int index)
        {
            list.RemoveAt(index);
            Count--;
        }

        /// <summary>
        /// Get all elements of list as array of syntax nodes
        /// </summary>
        /// <returns>Array of ISyntaxNode</returns>
        public ISyntaxNode[] GetElements()
        {
            return list.ToArray();
        }

        /// <summary>
        /// Clear the NodeList
        /// </summary>
        public new void Clear()
        {
            list.Clear();
            Count = 0;
        }

        /// <summary>
        /// Converts list to array
        /// </summary>
        /// <returns>Array of list</returns>
        public new ISyntaxNode[] ToArray()
        {
            return list.ToArray();
        }

        public override String ToString()
        {
            String text = "";
            ISyntaxNode[] nodes = list.ToArray();
            for (int i = 0; i <= (nodes.Length - 1); i++)
            {
                text += nodes[i].ToString();
            }

            return text;
        }

        public ISyntaxNode[] GetSubNodes()
        {
            return list.ToArray();
        }

        public void AcceptVisitor(ISyntaxNodeVisitor visitor)
        {
            visitor.Visit(this);
        }

        #endregion
    }
}

File: implementations/java/vanilla/src/main/java/org/cwi/waebric/parser/ast/NodeList.java
package org.cwi.waebric.parser.ast;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.ListIterator;


/**
 * Generic syntax node list implementation for syntax nodes that
 * represent a list structure.
 * 
 * @param <E>
 * 
 * @author Jeroen van Schagen
 * @date 20-05-2009
 */
public class NodeList<E extends AbstractSyntaxNode> extends AbstractSyntaxNode implements List<E> {

	/**
	 * Element collection
	 */
	protected List<E> list;
	
	/**
	 * Construct syntax node list.
	 */
	public NodeList() {
		list =  new ArrayList<E>();
	}

	/**
	 * Construct syntax node list.
	 * @param list Collection of all elements
	 */
	public NodeList(List<E> list) {
		this.list = list;
	}
	
	public boolean add(E element) {
		if(element == null) { return false; }
		if(this.contains(element)) { return false; }
		return list.add(element);
	}
	
	public void add(int index, E element) {
		list.add(index, element);
	}
	
	public boolean addAll(Collection<? extends E> c) {
		return list.addAll(c);
	}
	
	public boolean addAll(int index, Collection<? extends E> c) {
		return list.addAll(index, c);
	}
	
	public void clear() {
		list.clear();
	}
	
	@Override
	public NodeList<E> clone() {
		return new NodeList<E>(new ArrayList<E>(list));
	}
	
	public boolean contains(Object o) {
		return list.contains(o);
	}
	
	public boolean containsAll(Collection<?> c) {
		return list.containsAll(c);
	}
	
	@Override
	public boolean equals(Object arg) {
		if(arg instanceof NodeList) {
			NodeList<?> nodeList = (NodeList<?>) arg;
			if(nodeList.size() != this.size()) { return false; }
			for(int i = 0; i < this.size(); i++) {
				if(! this.get(i).equals(nodeList.get(i))) { return false; }
			}
			
			return true;
		}
		
		return false;
	}
	
	public E get(int index) {
		return list.get(index);
	}

	public AbstractSyntaxNode[] getChildren() {
		return list.toArray(new AbstractSyntaxNode[0]);
	}

	public int indexOf(Object o) {
		return list.indexOf(o);
	}

	public boolean isEmpty() {
		return list.isEmpty();
	}

	public Iterator<E> iterator() {
		return list.iterator();
	}

	public int lastIndexOf(Object o) {
		return list.lastIndexOf(o);
	}

	public ListIterator<E> listIterator() {
		return list.listIterator();
	}

	public ListIterator<E> listIterator(int index) {
		return list.listIterator();
	}

	public E remove(int index) {
		return list.remove(index);
	}

	public boolean remove(Object o) {
		return list.remove(o);
	}

	public boolean removeAll(Collection<?> c) {
		return list.retainAll(c);
	}

	public boolean retainAll(Collection<?> c) {
		return list.retainAll(c);
	}

	public E set(int index, E element) {
		return list.set(index, element);
	}

	public int size() {
		return list.size();
	}

	public List<E> subList(int fromIndex, int toIndex) {
		return list.subList(fromIndex, toIndex);
	}

	public Object[] toArray() {
		return list.toArray();
	}

	public <T> T[] toArray(T[] a) {
		return list.toArray(a);
	}
	
	@Override
	public String toString() {
		String name = "[";
		
		AbstractSyntaxNode[] children = this.getChildren();
		for(int i = 0; i < children.length; i++) {
			if(i != 0) { name += ","; }
			name += children[i].toString();
		}

		return name + "]";
	}
	
	@Override
	public void accept(INodeVisitor visitor, Object[] args) {
		visitor.visit(this, args);
	}

}